frontend/src/components/Search.js

```diff
@@ -0,0 +1,54 @@
+import React, { useState, useEffect } from "react";
+import { useHistory } from "react-router-dom";
+import { useDispatch } from "react-redux";
+import {
+  saveSearchedPosts,
+  saveSearchedQuery
+} from "../store/actions/actions";
+
+import { useRouteLeave } from "../utils/customHooks/onRouteLeave";
+
+const Search = () => {
+  const [input, setInput] = useState("");
+  const [timeoutID, setTimeoutID] = useState(null);
+  const dispatch = useDispatch();
+  const history = useHistory();
+  const hasLeavedRoute = useRouteLeave("/search");
+
+  useEffect(() => {
+    if (hasLeavedRoute) {
+      dispatch(saveSearchedQuery(""));
+      setInput("");
+    }
+  }, [hasLeavedRoute, dispatch]);
+
+  const handleInput = event => {
+    const targetValue = event.target.value;
+
+    setInput(targetValue);
+    clearTimeout(timeoutID);
+
+    dispatch(saveSearchedQuery(targetValue));
+
+    if (targetValue) {
+      const timeout = setTimeout(() => {
+        history.push("/search");
+        dispatch(saveSearchedPosts(targetValue));
+      }, 300);
```

Comment:
what will happen, if something take longer than 300

Reply:
It is a 'delay fetch data' mechanism, to avoid unnecessary api calls. 
If user stops typing for 300 ms, then 'fetch data' will be called. 
If not for the timeout, after every single letter there would be a call to api.